### PR TITLE
Change electron bundle loading strategy

### PR DESCRIPTION
**Pull request by chrisnojima**
Old way:

1. Single static renderer.(dev)html starts up and calls node to get a payload to load the scripts it needs
1. component-loader knows about all remote window types and operates off of an enum passed in

New way: 

1. Individual generated html and bundles for all windows (main, menubar, tracker, pinentry etc) directly loads their own scripts
1. component-loader gets injected with the data it needs and handles setting up the remote store generically

### CODE REVIEW

File: shared/README.md [hunk 1]
```diff
@@ -37,10 +37,7 @@ You can set environment variables for debugging:
 | Env     | Description |
 |---------|-------------|
 | KEYBASE_RUN_MODE | Run mode: prod, staging, devel |
-| KEYBASE_LOCAL_DEBUG | For debugging |
```
Comment (chrisnojima) [inline shared/README.md]: these aren't used anymore

File: shared/desktop/app/hot-path.desktop.js [hunk 2]
```diff
@@ -1,7 +0,0 @@
-// @flow
-import {resolveRoot} from './resolve-root.desktop'
-
-export default (path: string) => {
-  // $FlowIssue
-  return __HOT__ ? `http://localhost:4000/dist/${path}` : resolveRoot('dist', path) // eslint-disable-line no-undef
```
Comment (chrisnojima) [inline shared/desktop/app/hot-path.desktop.js]: none of this hot stuff needs to leak into the normal code anymore

File: shared/desktop/package.desktop.js [hunk 3]
```diff
@@ -194,7 +191,11 @@ function postPack(plat, arch) {
     }
     const subdir = plat === 'darwin' ? 'Keybase.app/Contents/Resources' : 'resources'
     const dir = path.join(appPaths[0], subdir, 'app/desktop/dist')
-    const files = ['index', 'main', 'component-loader'].map(p => p + '.bundle.js')
+    const modules = ['node', 'main', 'tracker', 'menubar', 'unlock-folders', 'pinentry']
+    const files = [
+      ...modules.map(p => p + '.bundle.js'),
+      ...modules.filter(p => p !== 'node').map(p => p + '.html'),
```
Comment (chrisnojima) [inline shared/desktop/package.desktop.js]: we have html files now too

File: shared/desktop/remote/component-loader.desktop.js [hunk 4]
```diff
@@ -1,106 +1,67 @@
...
 // This loads up a remote component. It makes a pass-through store which accepts its props from the main window through ipc
 // Also protects it with an error boundary
-import React, {Component} from 'react'
+import * as React from 'react'
+import * as SafeElectron from '../../util/safe-electron.desktop'
+import * as Styles from '../../styles'
 import ReactDOM from 'react-dom'
 import RemoteStore from './store.desktop'
 import Root from '../renderer/container.desktop'
-import Menubar from '../../menubar/remote-container.desktop'
```
Comment (chrisnojima) [inline shared/desktop/remote/component-loader.desktop.js]: this component used to know about everything and just switched between the remote types, now its more generic and expects stuff to be passed in

File: shared/desktop/renderer/index.html.template [hunk 5]
```diff
@@ -0,0 +1,26 @@
+<!DOCTYPE html>
+<html>
```
Comment (chrisnojima) [inline shared/desktop/renderer/index.html.template]: our template for the html files. based on the old hardcoded one

File: shared/desktop/webpack.config.babel.js [hunk 6]
```diff
@@ -160,24 +169,46 @@ const config = (_, {mode}) => {
...
+  const hmrPlugin = isHot && isDev ? [new webpack.HotModuleReplacementPlugin()] : []
+  const template = path.join(__dirname, './renderer/index.html.template')
+  const makeHtmlName = name => `${name}${isDev ? '.dev' : ''}.html`
+  const makeViewPlugins = names =>
+    [
+      ...hmrPlugin,
+      // Map since we generate multiple html files
+      ...names.map(
+        name =>
+          new HtmlWebpackPlugin({
```
Comment (chrisnojima) [inline shared/desktop/webpack.config.babel.js]: we now generate multiple html files that directly point to our js files

File: shared/desktop/webpack.config.babel.js [hunk 7]
```diff
@@ -160,24 +169,46 @@ const config = (_, {mode}) => {
...
+  }
+
+  // multiple entries so we can chunk shared parts
+  const entries = ['main', 'tracker', 'menubar', 'pinentry', 'unlock-folders']
+  const viewConfig = merge(commonConfig, {
+    entry: entries.reduce((map, name) => {
+      map[name] = `./${entryOverride[name] || name}/main.desktop.js`
+      return map
+    }, {}),
+    optimization: {splitChunks: {chunks: 'all'}},
```
Comment (chrisnojima) [inline shared/desktop/webpack.config.babel.js]: we now generate shared chunks so we don't have multiple copies of code between all our js bundles

File: shared/libs/flow-interface.js.flow [hunk 8]
```diff
@@ -19,327 +19,16 @@ declare module 'memoize-one' {
     equalityCheck?: (a: A1, b: A1) => boolean
   ): (A1, A2, A3, A4, A5, A6) => R
 }
-declare module '@babel/cli' {
```
Comment (chrisnojima) [inline shared/libs/flow-interface.js.flow]: a ton of these weren't used at all

File: shared/menubar/main.desktop.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+const style = {
+  ...Styles.globalStyles.flexBoxColumn,
+  borderTopLeftRadius: 4,
+  borderTopRightRadius: 4,
+  flex: 1,
+  marginTop: 0,
+  position: 'relative',
+}
+
+load({
```
Comment (chrisnojima) [inline shared/menubar/main.desktop.js]: each remote window has its own smaller bootstrap

File: shared/package.json [hunk 10]
```diff
@@ -225,6 +221,7 @@
     "flow-bin": "0.83.0",
     "flow-typed": "2.5.1",
     "fs-extra": "7.0.0",
+    "html-webpack-plugin": "3.2.0",
```
Comment (chrisnojima) [inline shared/package.json]: this is the only new thing, other stuff got auto sorted by yarn

File: shared/store/configure-sagas.js [hunk 11]
```diff
@@ -63,7 +63,7 @@ function create(crashHandler: (err: any) => void) {
 }
 
 function run() {
-  middleWare.run(mainSaga)
+  middleWare && middleWare.run(mainSaga)
```
Comment (chrisnojima) [inline shared/store/configure-sagas.js]: was crashing on hot reload w/o this

File: shared/unlock-folders/remote-container.desktop.js [hunk 12]
```diff
@@ -12,7 +12,7 @@ const mapDispatchToProps = dispatch => ({
   toPaperKeyInput: () => dispatch(UnlockFoldersGen.createToPaperKeyInput()),
 })
 export default connect(
-  state => state,
+  (state: any) => state,
```
Comment (chrisnojima) [inline shared/unlock-folders/remote-container.desktop.js]: flow is correct here, connect assumes the real store but the remote containers use a totally different store shape

File: shared/chat/conversation/messages/system-joined/container.js [hunk 13]
```diff
@@ -12,15 +12,17 @@ const mapStateToProps = (state, {message}) => ({
...
 })
 
-const mapDispatchToProps = (dispatch) => ({
+const mapDispatchToProps = dispatch => ({
   _onManageChannels: (teamname: string) =>
     isMobile
       ? dispatch(RouteTree.navigateTo([{props: {teamname}, selected: 'manageChannels'}], [chatTab]))
       : dispatch(RouteTree.navigateAppend([{props: {teamname}, selected: 'manageChannels'}])),
   onUsernameClicked: (username: string) => {
-    isMobile
```
Comment (buoyad) [inline shared/chat/conversation/messages/system-joined/container.js]: Could also just get rid of the brackets around this

**Comment by chrisnojima** [comment]
@keybase/react-hackers bump

**Comment by chrisnojima** [comment]
lemme see about the warnings. reload does work correctly

**Comment by chrisnojima** [comment]
@buoyad address both pr feedbacks. added a flag to never go through the main load() again

**Pull request merged**